# First-order properties

First-order properties quantify expectation values of one-electron operators in individual electronic states. For the electronic ground state, quantities such as the permanent electric dipole moment are evaluated directly from the converged Kohn–Sham density, requiring no response treatment. In contrast, expectation values for electronically excited states—most notably excited-state dipole moments—are obtained from the double residue of the quadratic response function, which provides the transition-specific way to construct expectation values in the excited-state manifold. Together, these formulations provide a consistent route to state-specific first-order properties across both ground and excited electronic states. See {cite}`Norman2018` for further details.

## Electric dipole moment

### Ground state



#### Simple translation test

In [31]:
import veloxchem as vlx

h2o_xyz = """3

O    10.000000000000        0.000000000000        0.000000000000
H    10.000000000000        0.740848095288        0.582094932012
H    10.000000000000       -0.740848095288        0.582094932012
"""

molecule = vlx.Molecule.read_xyz_string(h2o_xyz)
basis = vlx.MolecularBasis.read(molecule, "def2-svp")

scf_drv = vlx.ScfRestrictedDriver()
scf_drv.xcfun = "b3lyp"
scf_results = scf_drv.compute(molecule, basis)

prop_drv = vlx.FirstOrderPropertyDriver()

prop_drv.property = "electric dipole moment"

prop_results = prop_drv.compute(molecule, basis, scf_results)

                                                                                                                          
                                            Self Consistent Field Driver Setup                                            
                                                                                                                          
                   Wave Function Model             : Spin-Restricted Kohn-Sham                                            
                   Initial Guess Model             : Superposition of Atomic Densities                                    
                   Convergence Accelerator         : Two Level Direct Inversion of Iterative Subspace                     
                   Max. Number of Iterations       : 50                                                                   
                   Max. Number of Error Vectors    : 10                                                                   
                

In [32]:
print(prop_results)

{'electric dipole moment': {'total': array([ 1.95657334e-14, -9.36750677e-17,  7.80569421e-01]), 'nuclear': array([0., 0., 0.]), 'electronic': array([ 1.95657334e-14, -9.36750677e-17,  7.80569421e-01]), 'origin': array([18.89726125,  0.        ,  0.22      ]), 'units': 'a.u.'}, 'electric_dipole_moment': {'total': array([ 1.95657334e-14, -9.36750677e-17,  7.80569421e-01]), 'nuclear': array([0., 0., 0.]), 'electronic': array([ 1.95657334e-14, -9.36750677e-17,  7.80569421e-01]), 'origin': array([18.89726125,  0.        ,  0.22      ]), 'units': 'a.u.'}}


- It appears that results for key 'electric dipole moment' are entered twice into the dictionary.
- The origin is reported to be `(18.89726125, 0, 0.22)`. What does that mean?
- I am not getting any nuclear dipole moment although I have translated the water molecule in the x-direction

**Python script**

In [1]:
import veloxchem as vlx

molecule = vlx.Molecule.read_name("para-nitroaniline")
basis = vlx.MolecularBasis.read(molecule, "def2-svp")

scf_drv = vlx.ScfRestrictedDriver()
scf_drv.xcfun = "b3lyp"
scf_results = scf_drv.compute(molecule, basis)

prop_drv = vlx.FirstOrderPropertyDriver()

prop_drv.property = "electric dipole moment"

prop_results = prop_drv.compute(molecule, basis, scf_results)

Reading para-nitroaniline from PubChem...

Reference: S. Kim, J. Chen, T. Cheng, A. Gindulyte, J. He, S. He, Q. Li, B. A. Shoemaker, P. A. Thiessen, B. Yu, L. Zaslavsky, J. Zhang, E. E. Bolton, Nucleic Acids Res., 2025, 53, D1516-D1525.

Please double-check the compound since names may refer to more than one record.

                                                                                                                          
                                            Self Consistent Field Driver Setup                                            
                                                                                                                          
                   Wave Function Model             : Spin-Restricted Kohn-Sham                                            
                   Initial Guess Model             : Superposition of Atomic Densities                                    
                   Convergence Accelerator         : Two Level Dir

In [2]:
import numpy as np

results = prop_results["electric dipole moment"]

print("Electric dipole moment (a.u.):")
print(f"{"x":>17s} {"y":>8s} {"z":>8s}")

print(
    f"   nuclear: {results["nuclear"][0]:>8.4f} {results["nuclear"][1]:>8.4f} {results["nuclear"][2]:>8.4f}"
)
print(
    f"electronic: {results["electronic"][0]:>8.4f} {results["electronic"][1]:>8.4f} {results["electronic"][2]:>8.4f}"
)
print(
    f"     total: {results["total"][0]:>8.4f} {results["total"][1]:>8.4f} {results["total"][2]:>8.4f}"
)

dipmom = np.linalg.norm(results["total"])
au2debye = 2.5417464

print(f"\nMagnitude: {dipmom * au2debye: .4f} Debye")

Electric dipole moment (a.u.):
                x        y        z
   nuclear:  -0.0000   0.0000   0.0000
electronic:  -0.6933  -2.2896  -2.2007
     total:  -0.6933  -2.2896  -2.2007

Magnitude:  8.2621 Debye


In [3]:
molecule.show()

3Dmol.js failed to load for some reason. Please check your browser console for error messages.

### Excited state

**Python script**

In [ ]:
import veloxchem as vlx

molecule = vlx.Molecule.read_name("water")
basis = vlx.MolecularBasis.read(molecule, "def2-svp")

scf_drv = vlx.ScfRestrictedDriver()
scf_drv.xcfun = "cam-b3lyp"
scf_results = scf_drv.compute(molecule, basis)

In [4]:
from veloxchem.doubleresbeta import DoubleResBetaDriver

In [5]:
excmom_drv = DoubleResBetaDriver()

excmom_drv.initial_state = 3
excmom_drv.final_state = 3

excmom_results = excmom_drv.compute(molecule, basis, scf_results)

                                                                                                                          
                                             Quadratic Response Driver Setup                                              
                                                                                                                          
                                    ERI Screening Threshold         : 1.0e-12                                             
                                    Convergance Threshold           : 1.0e-04                                             
                                    Max. Number of Iterations       : 150                                                 
                                    Max. Number of Iterations       : 150                                                 
                                                                                                                          
                

In [6]:
excmom_results

{'photon_energies': [np.float64(-0.0)],
 'ground_state_dipole_moments': array([-0.69326284, -2.28958092, -2.20073492]),
 'excited_state_dipole_moments': {('x', 3, 3): np.float64(-0.4233535296736135),
  ('y', 3, 3): np.float64(-1.009678745860851),
  ('z', 3, 3): np.float64(-1.0207386840419561)},
 'oscillator_strengths': array([7.68671358e-10, 1.02693757e-04, 3.51936365e-01]),
 'elec_trans_dipoles': array([[-3.57845089e-05, -6.87736192e-05,  7.21898938e-05],
        [-1.04239821e-02, -2.03080268e-02,  2.44131310e-02],
        [-3.81741276e-01, -1.38797086e+00, -1.31765168e+00]]),
 'excitation_details': [['HOMO-1   -> LUMO         0.9909'],
  ['HOMO-4   -> LUMO         0.9899'],
  ['HOMO     -> LUMO         0.9916']]}

In [35]:
mu_g = np.array(excmom_results['ground_state_dipole_moments'])
print(f"\nmu_g: {np.linalg.norm(mu_g) * au2debye: .4f} Debye")

mu = []
for key, value in excmom_results['excited_state_dipole_moments'].items():
    mu.append(value)

mu = np.array(mu)

print(f"\nmu: {np.linalg.norm(mu) * au2debye: .4f} Debye")

mu_e = -(mu - 2 * mu_g)
print(f"\nmu_e: {np.linalg.norm(mu_e) * au2debye: .4f} Debye")

print(mu_e)


mu_g:  8.2621 Debye

mu:  3.8046 Debye

mu_e:  12.7337 Debye
[-0.96317215 -3.56948309 -3.38073115]


- It appears that what is calculated is $-(\mu_e - \mu_g)$ so we need to change sign before adding $\mu_g$
- What is computed is also the difference in electronic dipole moment so we need to add the nuclear part.
- polish: let us remove the plural "s" in the keys of the result dictionary

**Text file**

:::{code}
jobs
task: response
@end

@method settings
xcfun: cam-b3lyp
basis: def2-svpd
@end

@response
property: transition dipole moment
initial_state: 1
final_state: 1
@end

@molecule
charge: 0
multiplicity: 1
xyz:
...
@end
:::
